phpBB/includes/datetime.php

```diff
@@ -0,0 +1,158 @@
+<?php
+/**
+*
+* @package phpBB3
+* @copyright (c) 2012 phpBB Group
+* @license http://opensource.org/licenses/gpl-2.0.php GNU General Public License v2
+*/
+
+/**
+* phpBB custom extensions to the PHP DateTime class
+* This handles the relative formats phpBB employs
+*/
+class phpbb_datetime extends DateTime
+{
+	/**
+	* String used to wrap the date segment which should be replaced by today/tomorrow/yesterday
+	*/
+	const RELATIVE_WRAPPER = '|';
+
+	/**
+	* @var user User who is the context for this DateTime instance
+	*/
+	protected $user;
+
+	/**
+	* @var array Date formats are preprocessed by phpBB, to save constant recalculation they are cached.
+	*/
+	static protected $format_cache = array();
+
+	/**
+	* Constructs a new instance of phpbb_datetime, expanded to include an argument to inject
+	* the user context and modify the timezone to the users selected timezone if one is not set.
+	*
+	* @param string $time String in a format accepted by strtotime().
+	* @param DateTimeZone $timezone Time zone of the time.
+	* @param user User object for context.
+	*/
+	public function __construct($user, $time = 'now', DateTimeZone $timezone = null)
+	{
+		$this->user	= $user;
+		$timezone	= $timezone ?: $this->user->timezone;
+
+		parent::__construct($time, $timezone);
+	}
+
+	/**
+	* Formats the current date time into the specified format
+	*
+	* @param string $format Optional format to use for output, defaults to users chosen format
+	* @param boolean $force_absolute Force output of a non relative date
+	* @return string Formatted date time
+	*/
+	public function format($format = '', $force_absolute = false)
+	{
+		$format		= $format ? $format : $this->user->date_format;
+		$format		= self::format_cache($format, $this->user);
+		$relative	= ($format['is_short'] && !$force_absolute);
+		$now		= new self($this->user, 'now', $this->user->timezone);
+
+		$timestamp	= $this->getTimestamp();
+		$now_ts		= $now->getTimeStamp();
+
+		$delta		= $now_ts - $timestamp;
+
+		if ($relative)
+		{
+			/*
+			* Check the delta is less than or equal to 1 hour
+			* and the delta not more than a minute in the past
+			* and the delta is either greater than -5 seconds or timestamp
+			* and current time are of the same minute (they must be in the same hour already)
+			* finally check that relative dates are supported by the language pack
+			*/
+			if ($delta <= 3600 && $delta > -60 &&
+			  ($delta >= -5 || (($now_ts / 60) % 60) == (($timestamp / 60) % 60))
```

Comment:
@nickvergessen Hey Joas 😃 Sorry for disturbing after a long time, but do you probably remember the logic behind those `-5` and `-60` values to handle timestamps ahead from `now`. Cheers.

Reply:
Phew it's a long time, but I guess it was related to your computer time being out of sync which would otherwise should that to post will be in the future?

Reply:
Ah, so kind of server time misconfiguration in case it's within a minute with `now`? I'm just about patching this file for a small bug and was thinking about use of `DateInterval` class so trying to figure out what happens here :) Thanks.